sopel/irc/isupport.py

```diff
@@ -0,0 +1,380 @@
+    return (key, parser(value))
+
+
+class ISupport(object):
+    """Storage class for IRC's ``ISUPPORT`` feature.
+
+    An instance of ``ISupport`` can be used as a read-only dict, to store
+    feature advertised by the IRC server::
+
+        >>> isupport = ISupport(chanlimit=(('&', None), ('#', 70)))
+        >>> isupport['CHANLIMIT']
+        (('&', None) ('#', 70))
+        >>> isupport.CHANLIMIT  # some parameters are also properties
+        {
+            '&': None,
+            '#': 70,
+        }
+        >>> 'chanlimit' in isupport  # case-insensitive
+        True
+        >>> 'chanmode' in isupport
+        False
+        >>> isupport.CHANMODE  # not advertised by the server!
+        Traceback (most recent call last):
+          File "<stdin>", line 1, in <module>
+        AttributeError: 'ISupport' object has no attribute 'CHANMODE'
+
+    The list of possible parameters can be found at
+    `modern.ircdocs.horse's RPL_ISUPPORT Parameters`__.
+
+    .. __: https://modern.ircdocs.horse/#rplisupport-parameters
+    """
+    def __init__(self, **kwargs):
+        self.__isupport = dict(
+            (key.upper(), value)
+            for key, value in kwargs.items()
+            if not key.startswith('-'))
+
+    def __getitem__(self, key):
+        key_ci = key.upper()
+        if key_ci not in self.__isupport:
+            raise KeyError(key_ci)
+        return self.__isupport[key_ci]
+
+    def __contains__(self, key):
+        return key.upper() in self.__isupport
+
+    def __getattr__(self, name):
+        if name not in self.__isupport:
+            raise AttributeError(name)
+
+        return self.__isupport[name]
+
+    def __setattr__(self, name, value):
+        # make sure you can't set the value of any ISUPPORT attribute yourself
+        if name == '_ISupport__isupport':
+            # allow to set self.__isupport inside of the class
+            super(ISupport, self).__setattr__(name, value)
+        elif name in self.__isupport:
+            # reject any modification of __isupport
+            raise AttributeError("Can't set value for %r" % name)
+        elif name not in self.__dict__:
+            raise AttributeError('Unknown attribute')
+
+    def apply(self, **kwargs):
+        """Build a new instance of :class:`ISupport`.
+
+        :return: a new instance, updated with the latest advertised features
+        :rtype: :class:`ISupport`
+
+        This method applies the latest advertised features from the server:
+        the result contains the new and updated parameters, and doesn't contain
+        the removed parameters (marked by ``-{PARAMNAME}``)::
+
+            >>> updated = {'-AWAYLEN': None, 'NICKLEN': 25, 'CHANNELLEN': 10}
+            >>> new = isupport.apply(**updated)
+            >>> 'CHANNELLEN' in new
+            True
+            >>> 'AWAYLEN' in new
+            False
+
+        """
+        kwargs_ci = dict((key.upper(), value) for key, value in kwargs.items())
```

Comment:
Like the tests named with `_ci`, I think we can be clearer here.

```suggestion
        kwargs_upper = dict((key.upper(), value) for key, value in kwargs.items())
```